### PR TITLE
Handle VoteStateUpdates for outdated roots bigger than slots in existing VoteState

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
https://github.com/solana-labs/solana/issues/27361

#### Summary of Changes
Set root to largest slot in VoteState less than the proposed root

Fixes #

### CODE REVIEW

File: programs/vote/src/vote_state/mod.rs [hunk 1]
```diff
@@ -188,18 +188,35 @@ fn check_update_vote_state_slots_are_valid(
...
+                    .map(|current_root| lockout.slot > current_root)
+                    .unwrap_or(true);
+                // Ensure we're iterating from biggest to smallest vote in the
+                // current vote state
+                assert!(lockout.slot < prev_slot && is_bigger_than_root);
+                if lockout.slot <= new_proposed_root {
+                    vote_state_update.root = Some(lockout.slot);
+                    break;
+                }
+                prev_slot = lockout.slot;
```
Comment (AshwinSekar) [inline programs/vote/src/vote_state/mod.rs:211]: is the only purpose of `prev_slot` to check that we're iterating from biggest to smallest? Isn't the order of `vote_state` already checked elsewhere?

File: programs/vote/src/vote_state/mod.rs [hunk 2]
```diff
@@ -188,18 +188,35 @@ fn check_update_vote_state_slots_are_valid(
...
+            }
         }
     }
 
-    // index into the new proposed vote state's slots, starting with the root if it exists then
-    // we use this mutable root to fold the root slot case into this loop for performance
-    let mut check_root = vote_state_update.root;
+    // Index into the new proposed vote state's slots, starting with the root if it exists then
+    // we use this mutable root to fold checking the root slot into this the below loop
```
Comment (AshwinSekar) [inline programs/vote/src/vote_state/mod.rs:217]: -this : `we use this mutable root to fold checking the root slot into the below loop"`

File: programs/vote/src/vote_state/mod.rs [hunk 3]
```diff
@@ -2192,40 +2237,225 @@ mod tests {
...
+        );
+
+        // Test when `vote_state_update_root` is in `current_vote_state_slots` but it's not the latest
+        // slot and the `current_vote_state_root.is_some()`.
+        let earliest_slot_in_history = 5;
+        let current_vote_state_slots: Vec<Slot> = vec![1, 2, 3, 4];
+        let current_vote_state_root = Some(0);
+        let vote_state_update_slots_and_lockouts = vec![(5, 1)];
+        let vote_state_update_root = 4;
+        let expected_root = Some(4);
```
Comment (AshwinSekar) [inline programs/vote/src/vote_state/mod.rs:2337]: ❤️

**Review by AshwinSekar — APPROVED** [review]
some minor nits. looks good, nice tests.

File: programs/vote/src/vote_state/mod.rs (hunk 1 above)
Comment (carllin) [reply programs/vote/src/vote_state/mod.rs:211]: Yeah I think `vote_state_update` ordering is checked in `check_update_vote_state_slots_are_valid` and returns `SlotNotOrdered` if not, but the current `vote_state` itself was never checked I think

File: programs/vote/src/vote_state/mod.rs [hunk 4]
```diff
@@ -242,20 +266,31 @@ fn check_update_vote_state_slots_are_valid(
...
                         vote_state_update_indexes_to_filter.push(vote_state_update_index);
                     }
-                    if check_root.is_some() {
-                        // If the vote state update has a root < earliest_slot_hash_in_history
-                        // then we use the current root. The only case where this can happen
-                        // is if the current root itself is not in slot hashes.
-                        assert!(vote_state.root_slot.unwrap() < earliest_slot_hash_in_history);
-                        check_root = None;
+                    if let Some(new_proposed_root) = root_to_check {
+                        if is_root_fix_enabled {
```
Comment (carllin) [inline programs/vote/src/vote_state/mod.rs:278]: @AshwinSekar  added check here and simplified the asserts

File: programs/vote/src/vote_state/mod.rs [hunk 5]
```diff
@@ -188,18 +189,41 @@ fn check_update_vote_state_slots_are_valid(
...
+        .unwrap_or(false);
+
+    let original_proposed_root = vote_state_update.root;
+    if let Some(new_proposed_root) = original_proposed_root {
+        // If the new proposed root `R` is less than the earliest slot hash in the history
+        // such that we cannot verify whether the slot was actually was on this fork, set
+        // the root to the latest vote in the current vote that's less than R.
         if earliest_slot_hash_in_history > new_proposed_root {
             vote_state_update.root = vote_state.root_slot;
+            if is_root_fix_enabled {
```
Comment (carllin) [inline programs/vote/src/vote_state/mod.rs:203]: @AshwinSekar added check here as well

**Review by AshwinSekar — APPROVED** [review]
lgtm

**Pull request merged**